boards/BMS/Src/App/states/App_ChargeState.c

```diff
@@ -19,23 +28,47 @@ static void ChargeStateRunOnTick1Hz(struct StateMachine *const state_machine)
 
 static void ChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
 {
-    if (App_AllStatesRunOnTick100Hz(state_machine))
-    {
-        struct BmsWorld *         world   = App_SharedStateMachine_GetWorld(state_machine);
-        struct BmsCanTxInterface *can_tx  = App_BmsWorld_GetCanTx(world);
-        struct Charger *          charger = App_BmsWorld_GetCharger(world);
+    App_AllStatesRunOnTick100Hz(state_machine);
+
+    struct BmsWorld *         world       = App_SharedStateMachine_GetWorld(state_machine);
+    struct BmsCanTxInterface *can_tx      = App_BmsWorld_GetCanTx(world);
+    struct Charger *          charger     = App_BmsWorld_GetCharger(world);
+    struct Accumulator *      accumulator = App_BmsWorld_GetAccumulator(world);
 
+    // Enter the fault state once one of these conditions are true
+    // and indicate the reason for the transition over CAN:
+    // - Charger cable has disconnected
+    // - Charger has faulted
+    // - Maximum cell voltage reaches MAX_CELL_VOLTAGE_THRESHOLD V (Charging complete)
+    if (!App_Charger_IsConnected(charger) || App_Charger_HasFaulted(charger) ||
+        App_Accumulator_GetMaxVoltage(accumulator, NULL, NULL) > MAX_CELL_VOLTAGE_THRESHOLD)
+    {
         if (!App_Charger_IsConnected(charger))
         {
             App_CanTx_SetPeriodicSignal_CHARGER_DISCONNECTED_IN_CHARGE_STATE(can_tx, true);
-            App_SharedStateMachine_SetNextState(state_machine, App_GetFaultState());
         }
+        if (App_Charger_HasFaulted(charger))
+        {
+            App_CanTx_SetPeriodicSignal_CHARGER_FAULT_DETECTED(can_tx, true);
+        }
+        if (App_Accumulator_GetMaxVoltage(accumulator, NULL, NULL) > MAX_CELL_VOLTAGE_THRESHOLD)
```

Comment:
This is almost guaranteed to cause some sort of memory corruption. Gotta be careful when you pass in NULL as an arg